compiler/src/iree/compiler/Dialect/VMVX/Transforms/ResolveBufferDescriptors.cpp

```diff
@@ -0,0 +1,211 @@
+    if (sourceType.getRank() != subType.getRank()) {
+      return rewriter.notifyMatchFailure(op,
+                                         "rank reducing subview not supported");
+    }
+    int rank = sourceType.getRank();
+
+    // Create a descriptor for the source.
+    IndexType indexType = rewriter.getIndexType();
+    SmallVector<Type> sizeStrideTypes;
+    for (int i = 0; i < rank; i++) {
+      sizeStrideTypes.push_back(indexType);
+    }
+    auto sourceDesc = rewriter.create<IREE::VMVX::GetBufferDescriptorOp>(
+        loc, op.getBaseBuffer().getType(), indexType, sizeStrideTypes,
+        sizeStrideTypes, subview.getSource());
+
+    // For sizes, we just use the new ones, discarding the source.
+    SmallVector<Value> newSizes;
+    for (int i = 0; i < rank; ++i) {
+      if (subview.isDynamicSize(i)) {
+        newSizes.push_back(subview.getDynamicSize(i));
+      } else {
+        newSizes.push_back(constant(subview.getStaticSize(i)));
+      }
+      op.getSizes()[i].replaceAllUsesWith(newSizes.back());
+    }
+
+    // Apply stride multipliers.
+    SmallVector<Value> strides;
+    for (int i = 0; i < rank; ++i) {
+      Value currentStride;
+      if (subview.isDynamicStride(i)) {
+        currentStride = subview.getDynamicStride(i);
+      } else {
+        currentStride = constant(subview.getStaticStride(i));
+      }
+      currentStride = rewriter.createOrFold<arith::MulIOp>(
+          loc, sourceDesc.getStrides()[i], currentStride);
+      strides.push_back(currentStride);
+      op.getStrides()[i].replaceAllUsesWith(currentStride);
+    }
+
+    // Offsets.
+    Value offset = sourceDesc.getOffset();
+    for (int i = 0; i < rank; ++i) {
+      Value logicalOffset;
+      if (subview.isDynamicOffset(i)) {
+        logicalOffset = subview.getDynamicOffset(i);
+      } else {
+        int64_t staticOffset = subview.getStaticOffset(i);
+        if (staticOffset == 0) {
+          // Since added, just omit (will multiply to 0).
+          continue;
+        }
+        logicalOffset = constant(staticOffset);
+      }
+      Value physicalOffset =
+          rewriter.createOrFold<arith::MulIOp>(loc, logicalOffset, strides[i]);
+      offset = rewriter.create<arith::AddIOp>(loc, offset, physicalOffset);
+    }
+    op.getOffset().replaceAllUsesWith(offset);
+
+    // Base.
+    op.getBaseBuffer().replaceAllUsesWith(sourceDesc.getBaseBuffer());
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+struct FromHalInterfaceBindingSubspan
+    : public OpRewritePattern<GetBufferDescriptorOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(GetBufferDescriptorOp op,
+                                PatternRewriter &rewriter) const override {
+    auto binding =
+        op.getSource().getDefiningOp<IREE::HAL::InterfaceBindingSubspanOp>();
+    if (!binding) return failure();
+    auto loc = op.getLoc();
+    auto constant = [&](int64_t idxValue) -> Value {
```

Comment:
Going to recommend IndexSet again (and catch some of the uses below not using this function) - when adding ~dozen constants per op it can really bloat the IR.